### PR TITLE
Provide documentation for groovy_library

### PR DESCRIPTION
**Pull request by grumpyjames**
Test-plan:
I ran soy2html locally and it appears to produce reasonable output.

### CODE REVIEW

**Comment by grumpyjames** [comment]
A few caveats here:

- I have absolutely no idea what I'm doing, so most of the structure has been heavily borrowed from the existing doc.
- I've not tried too hard to share the content of the documentation between groovy_library and java_library yet.

File: docs/rule/groovy_library.soy [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    {param prettify: true /}
+    {param description}
+      A rule that is used to define a set of Groovy files that can be compiled together.
+
+      It can also be used to cross compile a set of Groovy and Java files.
+    {/param}
+    {param content}
+
+{call buck.rule}
+{param status: 'FROZEN' /}
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: I would not mark this as frozen now.

File: docs/rule/groovy_library.soy [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  Usually these will all end in <code>.groovy</code>, but if any of the files end in
+  <code>.java</code>, cross compilation using the jdk found in <code>JAVA_HOME</code> will occur.
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'resources' /}
+  {param default : '[]' /}
+  {param desc}
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+{/call}
+
+{call buck.arg}
+  {param name: 'deps' /}
+  {param default : '[]' /}
+  {param desc}
+  Rules (usually other <code>groovy_library</code> or <code>java_library</code> rules) that are
+  used to generate the classpath required to compile this <code>groovy_library</code>.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+{/call}
+
+{call buck.arg}
+  {param name: 'exported_deps' /}
+  {param default : '[]' /}
+  {param desc}
+  Other <code>groovy_library</code> and <code>java_library</code> rules that depend
+  on this rule will also include its <code>exported_deps</code> in their classpaths.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'provided_deps' /}
+  {param default : '[]' /}
+  {param desc}
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'source' /}
+  {param default : '<global value>' /}
+  {param desc}
+  Only used during cross compilation.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'target' /}
+  {param default : '<global value>' /}
+  {param desc}
+  Only used during cross compilation.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'java_version' /}
+  {param default: '<global value>' /}
+  {param desc}
+  Only used during cross compilation.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  {/param} 
+{/call}
+
+{call buck.arg}
+  {param name: 'extra_arguments' /}
+  {param default : '[]' /}
+  {param desc}
+  Only used during cross compilation.
+
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: `{call buck.java_library /}`

File: docs/rule/groovy_library.soy [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+{literal}<pre class="prettyprint lang-py">
+# A rule that compiles a single .groovy file.
+groovy_library(
+  name = 'example',
+  srcs = ['MySourceFile.groovy'],
+)
+
+# A rule that compiles all of the .groovy files under the directory in
+# which the rule is defined using glob()
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: It's better to do `<p>` before each example and have each one in its own box, IMO

**Comment by grumpyjames** [comment]
Ta for the review! Expect an improved version early next week.

File: docs/rule/groovy_library.soy (hunk 1 above)
Comment (grumpyjames) [inline docs/rule/groovy_library.soy]: In retrospect, that is well named enough that it should have actually triggered a response in my brain. Good catch, thanks.

File: docs/concept/buckconfig.soy [hunk 11]
```diff
@@ -944,6 +945,30 @@ $ buck targets --resolve-alias app#src_jar
...
+  {param name: 'groovy_home' /}
+  {param description}
+    This defines the value of <code>GROOVY_HOME</code> that Buck should use. If it is not provided,
+    Buck will use the system's <code>GROOVY_HOME</code> by default.
+  {/param}
+  {param raw_example}
+{literal}<pre class="prettyprint lang-ini">
+[groovy]
+  groovy_home = /path/to/groovy_home
+</pre>{/literal}
```
Comment (sdwilsh) [inline docs/concept/buckconfig.soy]: I missed this last time, but you don't need to use the `raw_example` macro here.  Just do `{param example: '/path/to/groovy_home /}`

File: docs/concept/buckconfig.soy [hunk 12]
```diff
@@ -944,6 +945,30 @@ $ buck targets --resolve-alias app#src_jar
...
+  {/param}
+{/call}
+
+
+
```
Comment (sdwilsh) [inline docs/concept/buckconfig.soy]: You only need one newline here :)

File: docs/rule/groovy_library.soy [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+  {param desc}
+  This is the same as in <a href="{ROOT}rule/java_library.html"><code>java_library</code></a>.
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'deps' /}
+  {param default : '[]' /}
+  {param desc}
+  Rules (usually other <code>groovy_library</code> or <code>java_library</code> rules) that are
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: Use `{call buck.java_library /}` here when mentioning `java_library` please

File: docs/rule/groovy_library.soy [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+  This is the same as in {call buck.java_library /}.
+  {/param}
+{/call}
+
+{call buck.arg}
+  {param name: 'exported_deps' /}
+  {param default : '[]' /}
+  {param desc}
+  Other <code>groovy_library</code> and <code>java_library</code> rules that depend
```
Comment (sdwilsh) [inline docs/rule/groovy_library.soy]: Use `{call buck.java_library /}` here when mentioning `java_library` please

**Comment by sdwilsh** [comment]
Nearly there!  You missed one comment from last time, plus I spotted a few more things that I missed.

**Comment by grumpyjames** [comment]
I think I got all of them this time. Thanks for the continuing scrutiny; it's clearly necessary :-)

File: docs/concept/buckconfig.soy [hunk 15]
```diff
@@ -944,6 +945,23 @@ $ buck targets --resolve-alias app#src_jar
...
+{/call}
+
+{call buckconfig.entry}
+  {param section: 'groovy' /}
+  {param name: 'groovy_home' /}
+  {param description}
+    This defines the value of <code>GROOVY_HOME</code> that Buck should use. If it is not provided,
+    Buck will use the system's <code>GROOVY_HOME</code> by default.
+  {/param}
+  {param example_value: '/path/to/groovy_home'/}
```
Comment (sdwilsh) [inline docs/concept/buckconfig.soy]: Can you plop this on top of the description please to be consistent with the rest of the file? :)

**Comment by sdwilsh** [comment]
You missed one minor thing :)

**Comment by sdwilsh** [comment]
@facebook-github-bot shipit

**Comment by sdwilsh** [comment]
I was looking at the commit for this while writing the release notes, and I noticed the email address you've been using isn't associated with your github account.  That may be intentional, but I thought I'd call it out :)

**Comment by grumpyjames** [comment]
Unintended consequence of moving between machines at work, I think; bit annoying to have left the inconsistency in the history though, sorry! Both addresses are relatively public so not a big deal; will prod the config of the checkout I'm using now back to the gmail account. Thanks for the heads up.

**Pull request closed without merging**